In [75]:
import pandas as pd
import numpy as np

## Load data from the "EFG FO" and "Ticker info" sheets

In [48]:
file_name = 'raw_data/EFG FO Data.xlsx'

# Load data from the "EFG FO" and "Ticker info" sheets.
df_efo = pd.read_excel(file_name, sheet_name = 0)
df_ticker_info = pd.read_excel(file_name, sheet_name = 1)

In [49]:
df_efo

Date       Ticker        FO WTD        FO MTD        FO YTD  \
0     2024-01-02  ADNOCGAS UH  8.425772e+00  8.425772e+00  8.425772e+00   
1     2024-01-02       DSI UH  0.000000e+00  0.000000e+00  0.000000e+00   
2     2024-01-02       UPP UH  4.764544e-11  4.764544e-11  4.764544e-11   
3     2024-01-02      WAHA UH -4.738458e-03 -4.738458e-03 -4.738458e-03   
4     2024-01-02   RAKBANK UH -3.313423e-02 -3.313423e-02 -3.313423e-02   
...          ...          ...           ...           ...           ...   
33815 2024-10-07       SPM AB -8.193077e-01 -1.468281e+00 -1.821806e+00   
33816 2024-10-07     NADEC AB -6.456229e-01 -1.088718e+00 -4.552456e+01   
33817 2024-10-07   MEDGULF AB -6.498437e-01 -1.047131e+00  1.635429e+01   
33818 2024-10-07       GIG AB  8.208959e-02 -4.167620e-02 -1.246699e+01   
33819 2024-10-07       KRE KK -1.054162e+00 -8.443315e-01  7.483948e+00   

       Foreign Headroom     FO%  
0               97.7331  2.0037  
1               27.0500  8.3600  
2               29.6900  4.3600  
3               42.5893  4.5679  
4               25.5851  1.3659  
...                 ...     ...  
33815           45.3200  3.6800  
33816           42.6400  6.3600  
33817           41.9000  7.1000  
33818           45.7400  3.2600  
33819           90.3600  9.6400  

[33820 rows x 7 columns]

In [50]:
df_ticker_info

Ticker       Country                  Sector      EXCHANGE
0    ALPHADHA UH           UAE             Industrials     Abu Dhabi
1     ALMARAI AB  SAUDI ARABIA        Consumer Staples  Saudi Arabia
2      ALINMA AB  SAUDI ARABIA                 Banking  Saudi Arabia
3         MIS AB  SAUDI ARABIA  Information Technology  Saudi Arabia
4        ALBI AB  SAUDI ARABIA                 Banking  Saudi Arabia
..           ...           ...                     ...           ...
283  ADNOCGAS UH           UAE                  Energy     Abu Dhabi
284  PUREHEAL UH           UAE             Health Care     Abu Dhabi
285    PARKIN UH           UAE             Industrials           DFM
286   ADNOCLS UH           UAE                  Energy     Abu Dhabi
287      ADES AB  SAUDI ARABIA                  Energy  Saudi Arabia

[288 rows x 4 columns]

## Inspect dfs

In [51]:
# Taking 5 samples
df_efo.sample(5)

Date       Ticker    FO WTD     FO MTD      FO YTD  \
16163 2024-05-01      BEMA QD  0.000000   0.000000    0.000000   
19022 2024-05-21  SOLUTION AB -4.672785 -21.560875 -102.561436   
13683 2024-04-16     ARCCI AB  0.851782  -6.844860   74.996243   
12924 2024-03-20     HERFY AB -0.123518   1.728145    3.775281   
29991 2024-09-11   EMPOWER UH  0.000000   0.917379   10.687927   

       Foreign Headroom   FO%  
16163           48.8696  0.00  
19022           42.8800  6.12  
13683           43.9200  5.08  
12924           44.3400  4.66  
29991           90.2300  9.64

281 unique values of `Ticker` column in df_efo

In [ ]:
df_efo['Ticker'].nunique()

281

In [53]:
df_efo['Ticker'].value_counts()

Ticker
ADNOCGAS UH    124
ALRAJHI AB     124
DERAYAHR AB    124
TADAWULG AB    124
WALAA AB       124
              ... 
ALMUTAHE KK     38
SMASCO AB       32
AZNOULA KK      24
MODON UH        16
NMDCENR UH      11
Name: count, Length: 281, dtype: int64

In [54]:
df_efo['Date'].value_counts()

Date
2024-09-29    277
2024-09-26    277
2024-09-25    277
2024-09-24    277
2024-09-23    277
             ... 
2024-03-11    269
2024-03-10    269
2024-03-07    269
2024-03-06    269
2024-03-19    269
Name: count, Length: 124, dtype: int64

281 unique values of `Ticker` column in df_ticker_info

In [55]:
df_ticker_info['Ticker'].nunique()

288

In [56]:
# Taking 5 samples
df_ticker_info.sample(5)

Ticker       Country       Sector      EXCHANGE
127     SAFCO AB  SAUDI ARABIA    Materials  Saudi Arabia
25       ARMX UH           UAE  Industrials           DFM
276  EMAARMLS UH           UAE  Real Estate           DFM
110   RAKBANK UH           UAE      Banking     Abu Dhabi
31       SECO AB  SAUDI ARABIA    Utilities  Saudi Arabia

In [57]:
df_ticker_info['Country'].value_counts()

Country
SAUDI ARABIA    155
UAE              64
QATAR            39
KUWAIT           30
Name: count, dtype: int64

In [58]:
df_ticker_info['Sector'].value_counts()

Sector
Industrials                        45
Materials                          42
Banking                            41
Real Estate                        30
Consumer Discretionary             26
Consumer Staples                   22
Financial Services                 15
Insurance                          13
Energy                             13
Communication Services             12
Health Care                        12
Utilities                          11
Information Technology              5
Consumer Discretionary Services     1
Name: count, dtype: int64

In [59]:
df_ticker_info['EXCHANGE'].value_counts()

EXCHANGE
Saudi Arabia    155
Qatar            39
Abu Dhabi        36
Kuwait           30
DFM              28
Name: count, dtype: int64

In [60]:
df_efo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33820 entries, 0 to 33819
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              33820 non-null  datetime64[ns]
 1   Ticker            33820 non-null  object        
 2   FO WTD            33820 non-null  float64       
 3   FO MTD            33792 non-null  float64       
 4   FO YTD            33820 non-null  float64       
 5   Foreign Headroom  33820 non-null  float64       
 6   FO%               33820 non-null  float64       
dtypes: datetime64[ns](1), float64(5), object(1)
memory usage: 1.8+ MB


#### `FO MTD` column has 28 null values

In [61]:
df_efo[df_efo['FO MTD'].isna()].head(5)

Date       Ticker    FO WTD  FO MTD    FO YTD  Foreign Headroom  \
319 2024-01-03   JAZEERA KK  0.000000     NaN  0.000000             83.50   
321 2024-01-03   MABANEE KK -0.356200     NaN -0.356200             84.15   
349 2024-01-03  INTEGRAT KK  0.041952     NaN  0.041952             84.81   
356 2024-01-03  WARBABAN KK  0.000000     NaN  0.000000             94.49   
357 2024-01-03  ALIMTIAZ KK -0.019628     NaN -0.019628             95.84   

       FO%  
319  16.50  
321  15.85  
349  15.19  
356   5.51  
357   4.16

- We can replace the nan values with the mean of the remaining 33792 values
- Or replace the nan values with the mean of only that day
- But lets drop those rows for now, hence they are only 28 values which represents only 0.082%

In [62]:
df_efo = df_efo[~df_efo['FO MTD'].isna()]
df_efo.info()


<class 'pandas.core.frame.DataFrame'>
Index: 33792 entries, 0 to 33819
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Date              33792 non-null  datetime64[ns]
 1   Ticker            33792 non-null  object        
 2   FO WTD            33792 non-null  float64       
 3   FO MTD            33792 non-null  float64       
 4   FO YTD            33792 non-null  float64       
 5   Foreign Headroom  33792 non-null  float64       
 6   FO%               33792 non-null  float64       
dtypes: datetime64[ns](1), float64(5), object(1)
memory usage: 2.1+ MB


In [63]:
df_efo['Ticker'].nunique()

281

In [64]:
df_ticker_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 288 entries, 0 to 287
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Ticker    288 non-null    object
 1   Country   288 non-null    object
 2   Sector    288 non-null    object
 3   EXCHANGE  288 non-null    object
dtypes: object(4)
memory usage: 9.1+ KB


In [65]:
df_efo.describe()

Date        FO WTD        FO MTD  \
count                          33792  33792.000000  3.379200e+04   
mean   2024-05-11 01:21:51.647727360      0.646435  2.445621e+00   
min              2024-01-02 00:00:00   -530.078761 -5.339566e+02   
25%              2024-02-26 00:00:00     -0.596962 -1.288028e+00   
50%              2024-05-07 00:00:00      0.000000  1.013679e-15   
75%              2024-08-01 00:00:00      0.591374  1.610163e+00   
max              2024-10-07 00:00:00   4715.709022  4.751768e+03   
std                              NaN     44.577025  7.242057e+01   

             FO YTD  Foreign Headroom           FO%  
count  33792.000000      33792.000000  33792.000000  
mean      16.981644         50.726312      9.267188  
min     -461.128525         -3.410000      0.000000  
25%       -4.281352         38.640000      4.090000  
50%        0.692450         43.460000      6.950000  
75%        9.836010         48.440000     11.610000  
max     4944.882900         99.940000     79.982200  
std      174.432982         21.661385      9.478095

In [66]:
df_ticker_info.describe()

Ticker       Country       Sector      EXCHANGE
count           288           288          288           288
unique          288             4           14             5
top     ALPHADHA UH  SAUDI ARABIA  Industrials  Saudi Arabia
freq              1           155           45           155

#### 281 unique values of `Ticker` column in df_efo while 288 values exists in df_ticker_info
This is the column on which the join will happen

In [69]:
df_ticker_info['Ticker'].nunique(), df_efo['Ticker'].nunique()

(288, 281)

There are 16 tickers that exists in df_ticker_info and not in df_efo

In [97]:
df_ticker_info_ticker_arr = df_ticker_info['Ticker'].unique()
df_efo_ticker_arr = df_efo['Ticker'].unique()

diff = np.setdiff1d(df_ticker_info_ticker_arr, df_efo_ticker_arr)
diff, diff.size

(array(['ALHOKAIR AB', 'ALKHODAR AB', 'ALMRAKEZ AB', 'ALQURAIN KK',
        'ARTC UH', 'AUB KK', 'AXA AB', 'DAMAC UH', 'DUR AB', 'DXBE UH',
        'EMAARMLS UH', 'ETISALAT UH', 'NCB AB', 'PETROCH AB', 'SAMBA AB',
        'SRE AB'], dtype=object),
 16)

In [98]:
df_ticker_info[df_ticker_info['Ticker'].isin(list(diff))]

Ticker       Country                  Sector      EXCHANGE
77   ALQURAIN KK        KUWAIT        Consumer Staples        Kuwait
84        AUB KK        KUWAIT                 Banking        Kuwait
116      ARTC UH           UAE             Industrials     Abu Dhabi
160  ETISALAT UH           UAE  Communication Services     Abu Dhabi
166     DAMAC UH           UAE             Real Estate           DFM
173       DUR AB  SAUDI ARABIA  Consumer Discretionary  Saudi Arabia
187  ALHOKAIR AB  SAUDI ARABIA  Consumer Discretionary  Saudi Arabia
188   PETROCH AB  SAUDI ARABIA               Materials  Saudi Arabia
194       AXA AB  SAUDI ARABIA               Insurance  Saudi Arabia
211  ALMRAKEZ AB  SAUDI ARABIA             Real Estate  Saudi Arabia
269       SRE AB  SAUDI ARABIA             Real Estate  Saudi Arabia
273       NCB AB  SAUDI ARABIA                 Banking  Saudi Arabia
274     SAMBA AB  SAUDI ARABIA                 Banking  Saudi Arabia
275  ALKHODAR AB  SAUDI ARABIA             Industrials  Saudi Arabia
276  EMAARMLS UH           UAE             Real Estate           DFM
277      DXBE UH           UAE  Consumer Discretionary           DFM

In [99]:
df_efo[df_efo['Ticker'].isin(list(diff))]

Empty DataFrame
Columns: [Date, Ticker, FO WTD, FO MTD, FO YTD, Foreign Headroom, FO%]
Index: []

There are 9 tickers that exists in df_efo and not in df_ticker_info

In [100]:
diff = np.setdiff1d(df_efo_ticker_arr, df_ticker_info_ticker_arr)
diff, diff.size

(array(['AGILITY UH', 'AVALONPH AB', 'EAT AB', 'MODON UH', 'NMDCENR UH',
        'SAL AB', 'SMASCO AB', 'SPINNEYS UH', 'SPM AB'], dtype=object),
 9)

In [101]:
df_efo[df_efo['Ticker'].isin(list(diff))]

Date      Ticker     FO WTD      FO MTD      FO YTD  \
17075 2024-05-08  AGILITY UH -19.741668  -43.254086  -43.254086   
17452 2024-05-09  AGILITY UH -25.423761  -48.750846  -48.750846   
17620 2024-05-12  AGILITY UH -29.018964  -51.812077  -51.812077   
17993 2024-05-13  AGILITY UH  -1.949558  -53.901054  -53.901054   
18165 2024-05-16  AGILITY UH -10.837825  -62.130164  -62.130164   
...          ...         ...        ...         ...         ...   
33706 2024-10-07  AGILITY UH  -1.412987   -2.675910 -252.713952   
33718 2024-10-07    MODON UH   5.387683 -524.461674 -271.860372   
33719 2024-10-07  NMDCENR UH   3.013885   65.726532  100.247603   
33797 2024-10-07      EAT AB  -1.627273   -4.389791   25.224555   
33815 2024-10-07      SPM AB  -0.819308   -1.468281   -1.821806   

       Foreign Headroom      FO%  
17075           51.4686  15.4538  
17452           51.4877  15.2995  
17620           51.5275  15.1929  
17993           51.5253  15.1392  
18165           51.6036  14.8932  
...                 ...      ...  
33706           51.7530   9.1997  
33718           27.3923   2.1862  
33719           91.4166   7.5929  
33797           40.6700   8.3300  
33815           45.3200   3.6800  

[392 rows x 7 columns]

Those 9 unique tickers results in 392 records, which make up about 1.1%

In [102]:
df_ticker_info[df_ticker_info['Ticker'].isin(list(diff))]

Empty DataFrame
Columns: [Ticker, Country, Sector, EXCHANGE]
Index: []

## Merge the two sheets on the "Ticker" column to create a single Data Frame